File: arch/eggroll/storage-service-cxx/src/LMDBStore.cc
/*
 * Copyright 2019 The FATE Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "LMDBStore.h"

using std::cout;
using std::endl;

LMDBStore::LMDBStore() {
}

LMDBStore::LMDBStore(const LMDBStore& other) {
    this->_dbDir = other._dbDir;
    this->storeInfo = other.storeInfo;
}

LMDBStore::~LMDBStore() {
    LOG(INFO) << "[LMDBStore::~LMDBStore] dbDir: " << _dbDir << ", desctructor use count: " << _env.use_count() << endl;
}

bool LMDBStore::init(string dataDir, StoreInfo& storeInfo) {
    std::stringstream ss;
    string delimiter = "/";
    bool result = true;
    std::exception_ptr eptr;
    try {
        string storeType = LMDB_TEMPORARY;
        if (boost::iequals(storeInfo.getStoreType(), LMDB)) {
            storeType = LMDB;
        }

        cout << storeInfo.toString() << endl;

        ss << dataDir
           << delimiter << storeType
           << delimiter << storeInfo.getNameSpace()
           << delimiter << storeInfo.getTableName()
           << delimiter << storeInfo.getFragment() << delimiter;

        ss >> this->_dbDir;

        cout << "[LMDBStore::init] dbDir: " << _dbDir << endl;
        LOG(INFO) << "[LMDBStore::init] dbDir: " << _dbDir << endl;

        boost::filesystem::path dst = this->_dbDir;
        boost::filesystem::create_directories(dst);

        cout << "[LMDBStore::init] ready to open: " << _dbDir << endl;

//        this->env.set_max_dbs(1).set_max_readers(256).set_mapsize(1UL * 1024UL * 1024UL * 1024UL);

        cout << "[LMDBStore::init] env set: " << _dbDir << endl;
//        this->env.open(dbDir.c_str(), 0, 0644);

        this->storeInfo = storeInfo;
        this->_env = getMDBEnv(_dbDir.data(), 0, 0644);
        this->_dbi = this->_env->openDB("main", MDB_CREATE);
        LOG(INFO) << "[LMDBStore::init] inited. dbdir: " << _dbDir << ", use_count: " << _env.use_count() << endl;
        cout << "[LMDBStore::init] inited. dbdir: " << _dbDir << ", use_count: " << _env.use_count() << endl;
    } catch (...) {
        eptr = std::current_exception();
        result = false;
    }
    handle_eptr(eptr, __FILE__, __LINE__, this->toString());

    return result;
}

void LMDBStore::put(const Operand *operand) {
    LOG(INFO) << "[LMDBStore::put] dbDir: " << _dbDir << endl;
    MDBRWTransaction rwtxn = _env->getRWTransaction();
    std::exception_ptr eptr;
    try {
        rwtxn.put(_dbi, operand->key(), operand->value());
        rwtxn.commit();
    } catch (...) {
        eptr = std::current_exception();
        rwtxn.abort();
    }
    handle_eptr(eptr, __FILE__, __LINE__, this->toString());
}

long LMDBStore::putAll(ServerReader<Operand> *reader) {
    LOG(INFO) << "[LMDBStore::putAll] dbDir: " << _dbDir << endl;
    MDBRWTransaction rwtxn = _env->getRWTransaction();
    long i = 0;

    std::exception_ptr eptr;
    try {
        Operand operand;

        long countInterval = 100000;
        long countRemaining = countInterval;

        while (reader->Read(&operand)) {
            rwtxn.put(_dbi, operand.key(), operand.value());
            /*
            cout << "key: " << operand.key().c_str() << ", size: " << operand.key().size() << ", strlen: " << strlen(operand.key().c_str())
                << "; value: " << operand.value().c_str() << ", size: " << operand.value().size() << ", strlen: " << strlen(operand.key().c_str()) << endl;
                */
            ++i;

            if (--countRemaining == 0) {
                cout << "batch count: " << i << ", at " << std::time(0) << endl;
                countRemaining = countInterval;
            }
        }

        rwtxn.commit();
        cout << "[LMDBStore::putAll] dbDir: " << _dbDir << ", total putAll: " << i << endl;
        LOG(INFO) << "[LMDBStore::putAll] dbDir: " << _dbDir << ", total putAll: " << i << endl;
    } catch (...) {
        eptr = std::current_exception();
        rwtxn.abort();
    }
    handle_eptr(eptr, __FILE__, __LINE__, this->toString());
    return i;
}

string_view LMDBStore::putIfAbsent(const Operand *operand) {
    LOG(INFO) << "[LMDBStore::putIfAbsent] dbDir: " << _dbDir << endl;
    MDBRWTransaction rwtxn = _env->getRWTransaction();
    string_view result;
    std::exception_ptr eptr;
    try {
        string_view key = operand->key();
        int rc = rwtxn.get(_dbi, key, result);
        if (MDB_NOTFOUND == rc) {
            rwtxn.put(_dbi, key, operand->value());
            result = operand->value();
        }
        cout << "putIfAbsent. rc: " << rc << ", result: " << result << endl;
        rwtxn.commit();
    } catch (...) {
        eptr = std::current_exception();
        rwtxn.abort();
    }
    handle_eptr(eptr, __FILE__, __LINE__, this->toString());

    return result;
}

string_view LMDBStore::delOne(const Operand *operand) {
    LOG(INFO) << "[LMDBStore::delOne] dbDir: " << _dbDir << endl;
    MDBRWTransaction rwtxn = _env->getRWTransaction();
    string_view oldValue;

    std::exception_ptr eptr;
    try {
        string_view key = operand->key();
        int rc = rwtxn.get(_dbi, key, oldValue);

        cout << "----------" << endl
             << "key to delete: " << key.data() << endl;

        if (MDB_NOTFOUND != rc) {
            rwtxn.del(_dbi, key);
        }
        rwtxn.commit();
        cout << "after delete: result: " << oldValue << endl;
    } catch (...) {
        eptr = std::current_exception();
        rwtxn.abort();
    }
    handle_eptr(eptr, __FILE__, __LINE__, this->toString());

    return oldValue;
}

bool LMDBStore::destroy() {
    LOG(INFO) << "[LMDBStore::destroy] dbDir: " << _dbDir << endl;
    bool result = false;
    std::exception_ptr eptr;
    size_t n;
    try {
        int env_use_count = _env.use_count();
        if (env_use_count > 1) {
            LOG(INFO) << "unable to destroy " << _dbDir << ". env use_count: " << env_use_count << endl;
            return false;
        }
        n = std::count(_dbDir.begin(), _dbDir.end(), '/');
        std::stringstream ss;
        string tableName;
        ss << this->storeInfo.getTableName();
        ss >> tableName;
        size_t tableNamePos = _dbDir.rfind(tableName);
        cout << "dbDir: " << _dbDir
        << ", storeInfo: " << this->storeInfo.toString()
        << ", tableName: " << tableName
        << ", tableNamePos: " << tableNamePos
        << ", size: " << _dbDir.size() << endl;

        if (n >= 4 && _dbDir.substr(0, 4) != "////") {
            //string dirToRemove = dbDir.substr(0, tableNamePos);
            string dirToRemove = _dbDir.substr(0, tableNamePos) + tableName;
            LOG(INFO) << "[LMDBStore::destroy] dirToRemove: " << dirToRemove << endl;
            cout << "dirToRemove: " << dirToRemove << endl;
            if (boost::filesystem::exists(dirToRemove) && boost::filesystem::is_directory(dirToRemove)) {
                boost::filesystem::remove_all(dirToRemove);
                result = true;
            }
        }
    } catch (...) {
        eptr = std::current_exception();
    }
    handle_eptr(eptr, __FILE__, __LINE__, this->toString());

    cout << "destroy: delimiter count: " << n << ", result: " << result << endl;

    return result;
}

long LMDBStore::count() {
    LOG(INFO) << "[LMDBStore::count] dbDir: " << _dbDir << endl;
    long result;
    MDBROTransaction rotxn = _env->getROTransaction();
    std::exception_ptr eptr;
    try {
        MDB_stat stat;
        mdb_stat(rotxn, _dbi, &stat);
        result = stat.ms_entries;

        // iterateAll();
        cout << "count: " << result << endl;
        LOG(INFO) << "count: " << result << endl;
    } catch (...) {
        eptr = std::current_exception();
    }
    handle_eptr(eptr, __FILE__, __LINE__, this->toString());

    return result;
}

string_view LMDBStore::get(const Operand *operand) {
    LOG(INFO) << "[LMDBStore::get] dbDir: " << _dbDir << endl;
    string_view result;
    MDBROTransaction rotxn = _env->getROTransaction();

    std::exception_ptr eptr;
    try {
        int found = rotxn.get(_dbi, operand->key(), result);
        cout << "found: " << found << ", result: " << result << endl;
    } catch (...) {
        eptr = std::current_exception();
    }
    handle_eptr(eptr, __FILE__, __LINE__, this->toString());

    // iterateAll();
    return result;
}

// (a, b]
void LMDBStore::iterate(const Range *range, ServerWriter<Operand> *writer) {
    LOG(INFO) << "[LMDBStore::iterate] dbDir: " << _dbDir << endl;
    MDBROTransaction rotxn = _env->getROTransaction();
    MDBROCursor rocursor = rotxn.getCursor(_dbi);
    int count = 0;

    std::exception_ptr eptr;
    try {
        long bytesCount = 0;

        string start = range->start();
        string end = range->end();
        long threshold = range->minchunksize() > 0 ? range->minchunksize() : PAYLOAD_THREASHOLD;
        cout << "start: " << start << " (" << start.size() << "), end: " << end << ", threshold: " << threshold << endl;

        MDBOutVal key, val;
        string_view keyView, valView;

        bool locateStart = true;
        if (start.empty()) {
            locateStart = false;
        }

        cout << "locateStart: " << locateStart << start.empty() << start.size() << endl;
        bool checkEnd = true;
        if (end.empty()) {
            checkEnd = false;
        }

        int rc;
        Operand operand;

        // first element
        if (locateStart) {
            // key == actual start returned by the call
            rc = rocursor.lower_bound(start, key, val);
            keyView = key.get<string_view>();
            valView = val.get<string_view>();

            cout << "start: " << start <<
                 ", actual start: " << keyView
                 << ", isStartFound: " << (start == keyView);
            cout << endl;

            // first element
            if (keyView != start) {
                if (checkEnd && keyView.compare(end) > 0) {
                    return;
                }
                operand.set_key(keyView.data(), keyView.size());
                operand.set_value(valView.data(), valView.size());
                writer->Write(operand);
                ++count;

                bytesCount += keyView.size() + valView.size();

/*                cout << "key size: " << key.size() << ", key data: " << curKey << ", value size: " << value.size()
                     << ", value data: " << value.to_string_view() << endl;*/
            }
        }

        cout << "located start: " << keyView << endl;

        // 2 to last
        while (0 == rocursor.next(key, val)) {
            if (bytesCount >= threshold) {
                break;
            }

            keyView = key.get<string_view>();
            valView = val.get<string_view>();

            if (checkEnd && keyView.compare(end) > 0) {
                cout << "breaking. curKey: " << keyView << ", end: " << end << endl;
                break;
            }

            operand.set_key(keyView.data(), keyView.size());
            operand.set_value(valView.data(), valView.size());

            writer->Write(operand);

/*            cout << "key size: " << key.size() << ", key data: " << curKey << ", value size: " << value.size()
                 << ", value data: " << value.to_string() << endl;*/

            ++count;
            bytesCount += keyView.size() + valView.size();
        }
        cout << "[LMDBStore::iterate] dbDir: " << _dbDir << ", total iterated: " << count << endl;
        LOG(INFO) << "[LMDBStore::iterate] dbDir: " << _dbDir << ", total iterated: " << count << endl;
    } catch (...) {
        eptr = std::current_exception();
    }

    handle_eptr(eptr, __FILE__, __LINE__, this->toString());
}

void LMDBStore::iterateAll() {
    MDBROTransaction rotxn = _env->getROTransaction();
    MDBROCursor rocursor = rotxn.getCursor(_dbi);

    cout << "------ iterateAll ------" << endl;

    std::exception_ptr eptr;
    try {

        MDBOutVal key, val;

        int count = 0;

          while(!rocursor.get(key, val, count ? MDB_NEXT : MDB_FIRST)) {
            cout << key.get<string>();
            cout<<": " << val.get<string>();
            cout << "\n";
            ++count;

          }

        cout << endl << "------ iterateAll total: " << count << " ------" << endl;

        int rc;
        if (!(rc = rocursor.find("it", key, val))) {
            string_view keyResult(key.get<string_view>());
            cout << keyResult << ": " << val.get<string_view>() << endl;
        } else {
            cout << "rc: " << rc << endl;
        }
    } catch (...) {
        eptr = std::current_exception();
    }
    handle_eptr(eptr, __FILE__, __LINE__, this->toString());
}

string LMDBStore::toString() {
    std::stringstream ss;
    std::exception_ptr eptr;
    try {
        ss << "{storeInfo: " << this->storeInfo.toString()
           << ", dbDir: " << this->_dbDir
           << "}";
    } catch (...) {
        eptr = std::current_exception();
    }
    handle_eptr(eptr, __FILE__, __LINE__, "error in LMDBStore::toString()");

    string result;
    ss >> result;

    return result;
}

File: eggroll/storage/storage-service-java/src/main/java/com/webank/ai/fate/eggroll/storage/service/model/LMDBStore.java
/*
 * Copyright 2019 The FATE Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.webank.ai.fate.eggroll.storage.service.model;

import com.webank.ai.fate.core.error.exception.InvalidStateStoreException;
import com.webank.ai.fate.core.error.exception.ProcessorStateException;
import com.webank.ai.fate.core.io.KeyValue;
import com.webank.ai.fate.core.io.KeyValueIterator;
import com.webank.ai.fate.core.io.KeyValueStore;
import com.webank.ai.fate.core.io.StoreInfo;
import com.webank.ai.fate.core.model.Bytes;
import com.webank.ai.fate.core.utils.AbstractIterator;
import com.webank.ai.fate.core.utils.ErrorUtils;
import io.grpc.stub.StreamObserver;
import org.apache.commons.io.FileUtils;
import org.apache.commons.lang3.exception.ExceptionUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.iq80.leveldb.DBException;
import org.lmdbjava.*;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;


public class LMDBStore implements KeyValueStore<Bytes, byte[]> {
    public static final String DATA_DIR = "data.dir";
    private static final String TOSTRING_FORMAT = "LMDBStore : %s";
    private static Logger LOGGER = LogManager.getLogger(LMDBStore.class);
    private final StoreInfo storeInfo;
    private final Set<KeyValueIterator> openIterators = Collections.synchronizedSet(new HashSet<KeyValueIterator>());
    private File dbDir;
    private volatile boolean open = false;
    private Env<byte[]> env;
    private Dbi<byte[]> dbi;

    private ErrorUtils errorUtils;

    public LMDBStore(StoreInfo info) {
        this.storeInfo = info;
        this.errorUtils = new ErrorUtils();
    }

    @Override
    public void put(Bytes key, byte[] value) {
        Objects.requireNonNull(key, "key cannot be null");
        if (value == null || value.length == 0) {
            dbi.delete(key.get());
        } else {
            dbi.put(key.get(), value);
        }
    }

    @Override
    public byte[] putIfAbsent(Bytes key, byte[] value) {
        Objects.requireNonNull(key, "key cannot be null");

        try (Txn<byte[]> txn = env.txnWrite()) {
            byte[] oldValue = dbi.get(txn, key.get());
            if (oldValue == null) {
                if (value == null || value.length == 0) {
                    dbi.delete(txn, key.get());
                } else {
                    dbi.put(txn, key.get(), value);
                }
            }
            txn.commit();
            return oldValue;
        }
    }

    @Override
    public void putAll(List<KeyValue<Bytes, byte[]>> entries) {
        try (Txn<byte[]> txn = env.txnWrite()) {
            for (KeyValue<Bytes, byte[]> entry : entries) {
                Objects.requireNonNull(entry.key, "key cannot be null");
                if (entry.value == null) {
                    dbi.delete(txn, entry.key.get());
                } else {
                    dbi.put(txn, entry.key.get(), entry.value);
                }
            }
            txn.commit();
        }
    }

    @Override
    public StreamObserver<KeyValue<Bytes, byte[]>> putAll() {
        Txn<byte[]> txn = env.txnWrite();
        return new StreamObserver<KeyValue<Bytes, byte[]>>() {
            @Override
            public void onNext(KeyValue<Bytes, byte[]> entry) {
                Objects.requireNonNull(entry.key, "key cannot be null");
                if (entry.value == null) {
                    dbi.delete(txn, entry.key.get());
                } else {
                    dbi.put(txn, entry.key.get(), entry.value);
                }
            }

            @Override
            public void onError(Throwable throwable) {
                txn.abort();
                txn.close();
                LOGGER.error(errorUtils.toGrpcRuntimeException(throwable));
                LOGGER.info("[STORAGESERVICE][PUTALL] error");
            }

            @Override
            public void onCompleted() {
                synchronized (this) {
                    txn.commit();
                    txn.close();
                }
                //env.sync(true);
                LOGGER.info("[STORAGESERVICE][STORE][PUTALL] completed. store info: {}");
            }
        };
    }

    @Override
    public byte[] delete(Bytes key) {
        Objects.requireNonNull(key, "key cannot be null");
        try (Txn<byte[]> txn = env.txnWrite()) {
            byte[] value = dbi.get(txn, key.get());
            dbi.delete(txn, key.get());
            txn.commit();
            return value;
        }
    }

    @Override
    public byte[] get(Bytes key) {
        Objects.requireNonNull(key, "key cannot be null");
        try (Txn<byte[]> txn = env.txnRead()) {
            return dbi.get(txn, key.get());
        }
    }

    @Override
    public KeyValueIterator<Bytes, byte[]> range(Bytes from, Bytes to) {
        validateStoreOpen();
        final LMDBRangeIterator lmdbRangeIterator = new LMDBRangeIterator(from, to);
        openIterators.add(lmdbRangeIterator);
        return lmdbRangeIterator;
    }

    @Override
    public KeyValueIterator<Bytes, byte[]> all() {
        validateStoreOpen();
        final LMDBRangeIterator lmdbIterator = new LMDBRangeIterator(null, null);
        openIterators.add(lmdbIterator);
        return lmdbIterator;
    }

    @Override
    public void destroy() {
        try {
            try (Txn<byte[]> txn = env.txnWrite()) {
                dbi.drop(txn, true);
            }
            String[] files = dbDir.list();
            if (null != files) {
                for (String s : files) {
                    File currentFile = new File(dbDir.getPath(), s);
                    FileUtils.deleteQuietly(currentFile);
                }
            }
        } catch (Exception e) {
            LOGGER.info("[STORAGE] error in destroy: " + ExceptionUtils.getStackTrace(e));
            Thread.currentThread().interrupt();
            throw new RuntimeException(e);
        }
        this.close();
    }


    @Override
    public String name() {
        return storeInfo.getTableName();
    }

    @Override
    public long count() {
        return env.stat().entries;
    }

    private void validateStoreOpen() {
        if (!open) {
            throw new InvalidStateStoreException("Store " + storeInfo + " is currently closed");
        }
    }

    @Override
    public void init(Properties properties) {

        Path dbPath = Paths.get(properties.getProperty(DATA_DIR),
                storeInfo.getNameSpace(),
                storeInfo.getTableName(),
                storeInfo.getFragment() + "");

        try {
            try {
                Files.createDirectories(dbPath);
                dbDir = dbPath.toFile();
                env = Env.create(new ByteArrayProxy()).setMaxDbs(1).setMaxReaders(256).setMapSize(1_073_741_824).open(dbDir, EnvFlags.MDB_NOTLS, EnvFlags.MDB_NOSYNC, EnvFlags.MDB_NOLOCK);
                dbi = env.openDbi((String) null, DbiFlags.MDB_CREATE);
            } catch (final DBException e) {
                throw new ProcessorStateException("Error opening store " + storeInfo + " at location " + dbDir.toString(), e);
            }
        } catch (final IOException e) {
            throw new ProcessorStateException(e);
        }

        open = true;
    }

    @Override
    public void flush() {

    }

    @Override
    public void close() {
        if (!open) {
            return;
        }

        this.open = false;
        closeOpenIterators();
        env.close();
        dbi.close();
    }

    private void closeOpenIterators() {
        final HashSet<KeyValueIterator> iterators;
        synchronized (openIterators) {
            iterators = new HashSet<>(openIterators);
        }
        for (final KeyValueIterator iterator : iterators) {
            iterator.close();
        }
    }

    @Override
    public boolean persistent() {
        return true;
    }

    @Override
    public boolean isOpen() {
        return this.open;
    }

    @Override
    public String toString() {
        return String.format(TOSTRING_FORMAT, storeInfo.toString());
    }

    private class LMDBRangeIterator extends AbstractIterator<KeyValue<Bytes, byte[]>> implements KeyValueIterator<Bytes, byte[]> {
        final CursorIterator<byte[]> cursorIterator;
        final Txn<byte[]> txn;

        private volatile boolean open = true;
        private KeyValue<Bytes, byte[]> next;

        LMDBRangeIterator(final Bytes from, final Bytes to) {
            KeyRange<byte[]> keyRange;
            if (from != null && to != null) {
                keyRange = KeyRange.open(from.get(), to.get());
            } else if (from != null) {
                keyRange = KeyRange.greaterThan(from.get());
            } else if (to != null) {
                keyRange = KeyRange.lessThan(to.get());
            } else {
                keyRange = KeyRange.all();
            }
            this.txn = env.txnRead();
            this.cursorIterator = dbi.iterate(txn, keyRange, Bytes.BYTES_LEXICO_COMPARATOR);
        }

        @Override
        public synchronized void close() {
            openIterators.remove(this);
            txn.close();
            cursorIterator.close();
            this.open = false;
        }

        @Override
        public synchronized boolean hasNext() {
            if (!open) {
                throw new InvalidStateStoreException(String.format("LMDB store %s has closed", storeInfo.toString()));
            }
            return super.hasNext();
        }

        @Override
        public Bytes peekNextKey() {
            if (!hasNext()) {
                throw new NoSuchElementException();
            }
            return next.key;
        }

        @Override
        public KeyValue<Bytes, byte[]> allDone() {
            KeyValue<Bytes, byte[]> rtn = super.allDone();
            close();
            return rtn;
        }

        @Override
        public synchronized KeyValue<Bytes, byte[]> next() {
            return super.next();
        }

        @Override
        protected KeyValue<Bytes, byte[]> makeNext() {
            if (!cursorIterator.hasNext()) {
                return allDone();
            } else {
                CursorIterator.KeyVal<byte[]> keyVal = cursorIterator.next();
                next = new KeyValue<>(Bytes.wrap(keyVal.key()), keyVal.val());
                return next;
            }
        }
    }


}